src/rs/_internal/canonical_fsm/canonical_fsm.rs

```diff
@@ -160,65 +150,77 @@ impl<TPuzzle: SemiGroupActionPuzzle> CanonicalFSM<TPuzzle> {
 
         let mut queue_index: CanonicalFSMState = CANONICAL_FSM_START_STATE;
-        while Into::<usize>::into(queue_index) < state_to_mask.0.len() {
+        while usize::from(queue_index) < state_to_mask.0.len() {
             let mut next_state: IndexedVec<MoveClassIndex, CanonicalFSMState> =
                 IndexedVec::new(vec![ILLEGAL_FSM_STATE; num_move_classes]);
 
-            let dequeue_move_class_mask: MoveClassMask = state_to_mask.get(queue_index);
-            disallowed_move_classes.push(MoveClassMask(0));
+            let dequeue_move_class_mask = state_to_mask.get(queue_index).clone();
 
             queue_index += CanonicalFSMState(1);
-            let from_state = queue_index;
 
             for move_class_index in generators.by_move_class.index_iter() {
                 let mut skip = false;
                 // If there's a greater move (multiple) in the state that
                 // commutes with this move's `move_class`, we can't move
                 // `move_class`.
-                skip |= (*dequeue_move_class_mask & *commutes[*move_class_index])
-                    >> (*move_class_index + 1)
-                    != 0;
-                skip |= (*dequeue_move_class_mask & *forbidden_transitions[*move_class_index]) != 0;
-                skip |= ((*dequeue_move_class_mask >> *move_class_index) & 1) != 0;
+                skip |= dequeue_move_class_mask
+                    .0
+                    .iter()
+                    .zip(commutes[*move_class_index].0.iter())
+                    .skip(*move_class_index + 1)
+                    .any(|(&dequeue_move_class, &commute)| dequeue_move_class && commute);
+                skip |= dequeue_move_class_mask
+                    .0
+                    .iter()
+                    .zip(forbidden_transitions[*move_class_index].0.iter())
+                    .any(|(&dequeue_move_class, &forbidden_transition)| {
+                        dequeue_move_class && forbidden_transition
+                    });
+                skip |= dequeue_move_class_mask.0[*move_class_index];
                 if skip {
-                    let new_value = MoveClassMask(
-                        *disallowed_move_classes.get(from_state) | (1 << *move_class_index),
-                    );
-                    disallowed_move_classes.set(from_state, new_value);
                     continue;
                 }
-                // TODO implement bit arithmetic for whole number newtypes.
-                let mut next_state_bits = (*dequeue_move_class_mask & *commutes[*move_class_index])
-                    | (1 << *move_class_index);
+                let mut next_state_mask = dequeue_move_class_mask.0.clone();
+                for (next_state, commute) in next_state_mask
+                    .iter_mut()
+                    .zip(&commutes[*move_class_index].0)
+                {
+                    // TODO implement bit arithmetic for whole number newtypes.
```

Comment:
I think this should be able to use a simple `|=` if `next_state_mask` is a `MoveClassMask `?